omegaconf/basecontainer.py

```diff
@@ -798,3 +815,27 @@ def _update_types(node: Node, ref_type: type, object_type: Optional[type]) -> No
         if new_ref_type is not Any:
             node._metadata.ref_type = new_ref_type
             node._metadata.optional = new_is_optional
+
+
+def _instantiate_structured_config_impl(
+    retdict: Dict[str, Any], object_type: Type[Any]
+) -> Any:
+    from ._utils import get_structured_config_data
+
+    object_type_field_names = get_structured_config_data(object_type).keys()
+    if issubclass(object_type, dict):
+        # Extending dict as a subclass
+
+        retdict_field_items = {
+            k: v for k, v in retdict.items() if k in object_type_field_names
+        }
+        retdict_nonfield_items = {
+            k: v for k, v in retdict.items() if k not in object_type_field_names
+        }
+        result = object_type(**retdict_field_items)
+        result.update(retdict_nonfield_items)
+    else:
+        # normal structured config
+        assert set(retdict.keys()) <= set(object_type_field_names)
```

Comment:
Does this handle the case where the user add an additional field to the Structurec Config that is not in the underlying type?

Reply:
It does not handle that case. What is the desired behavior when there are extra fields: Do we want to call `setattr` on the dataclass instance?

Reply:
What is the purpose of the assert?


> What is the desired behavior when there are extra fields: Do we want to call setattr on the dataclass instance?

```python
In [1]: from dataclasses import *
   ...: from typing import *
   ...: from omegaconf import OmegaConf
   ...: @dataclass
   ...: class Foo:
   ...:     x: int
   ...: foo = Foo(x=10)

In [2]: foo
Out[2]: Foo(x=10)

In [3]: setattr(foo, "y", 20)

In [4]: foo
Out[4]: Foo(x=10)

In [5]: foo.y
Out[5]: 20
```

I actually had no idea this was possible.
Yes, I think it would be nice to use setattr here.
Other alternatives are to error out, or  to outright ignore the extra fields. But I think setattr is matching the OmegaConf behavior very well so we can try it.

Reply:
Ok, sounds good!

Reply:
> What is the purpose of the assert?

Sorry that I missed this.

If object_type is `MyDataclass`, we want to make sure that calling `MyDataclass(k1=v1, k2=v2, ...)` will not raise a TypeError.

Thus, before calling `MyDataclass(**instance_data)`, we assert that `instance_data.keys()` is a subset of the keywords arguments that `MyDataclass.__init__` is able to accept.

Reply:
Commit https://github.com/omry/omegaconf/pull/502/commits/62f34cf190f4ae6cb439a05cabcf7b1cbac791a3 implements `setattr(instance, k, v)` for extra fields appearing in a structured config.

Reply:
> Commit [62f34cf](https://github.com/omry/omegaconf/commit/62f34cf190f4ae6cb439a05cabcf7b1cbac791a3) implements `setattr(instance, k, v)` for extra fields appearing in a structured config.

What's the motivation? The assert made sense to me: I thought only "dict-like" dataclasses were allowed to be extended with new fields?

Reply:
As I see it, there are both pros and cons to this `setattr` behavior.

The major cons are:
1. type safety is reduced, and
2. accidental calls to `UserDataclass.__setattr__` can have serious consequences:
  - if `UserDataclass` has any methods, those methods could be overwritten by
    our calling `setattr(instance, 'methodname', value)`
  - if `UserDataclass` implements a custom `__setattr__` method, our calling
    `setattr(instance, 'key', value)` could trigger side effects.

The main pro is:

OmegaConf and Hydra both make it easy for clients to add extra fields to
Structured Configs, using `omegaconf.open_dict` and the Hydra `+key=value` CLI
syntax, respectively. Not implementing the `setattr` behavior would mean that
users are restricted from using `open_dict`/`+key=value` in combination with
`to_object` when structured configs are involved.

I'd like to propose the following: enforce the schema strictly by default
(raising an error if there are extra fields), and allow users to opt in to the
`setattr` behavior. For example, we could add another member to the `SCMode`
Enum, say `SCMode.INSTANTIATE_SETATTR`, so that users can opt in to the
behavior by selecting this member (with `SCMode.INSTANTIATE` (no `setattr`)
being the default).

Reply:
Here is some additional motivation for allowing users to opt in to `setattr`
behavior:

For Python objects in general (not just for dataclasses), the default
implementation of `__setattr__` allows for flexible attribute assignment:
```python
class A:
    pass

A.x = 123  # A.x becomes a class variable
a = A()
a.y = 456  # A.y becomes an instance variable
assert a.x == 123
assert a.y == 456
```
This is pythonic; it works because `A` is a subclass of the `object` class and
`object.__setattr__` implements this behavior. Users can take advantage of this
to achieve something similar to the `namespace` from C++.

Here is an example of how this might look with OmegaConf:
```python
@dataclass
class MyNamespace:
    required_id: str

    @dataclass
    class SubNamespace:
        pass

    sub1: SubNamespace = field(default_factory=SubNamespace)
    sub2: SubNamespace = field(default_factory=SubNamespace)

cfg = OmegaConf.structured(MyNamespace)
cfg.required_id = "bond"
with open_dict(cfg):
    cfg.foo = "bar"

with open_dict(cfg.sub1):
    cfg.sub1.baz = "qux"

data: MyNamespace = OmegaConf.to_object(cfg, structured_config_mode=SCMode.INSTANTIATE_SETATTR)

assert data.foo == "bar"
assert data.sub1.baz == "qux"
```

Reply:
> OmegaConf and Hydra both make it easy for clients to add extra fields to Structured Configs, using `omegaconf.open_dict` and the Hydra `+key=value` CLI syntax, respectively.

Just that is enough to convince me. However, in that case I don't think we should treat differently a dataclass that's been extended through `open_dict()` vs a dataclass that is declared "extendable" by inheriting from Dict => I would suggest to make all extra fields new attributes in `to_object()`.

Reply:
> Edit: actually, in addition to that, I think that if the dataclass inherits from Dict then the instance returned by `to_object()` should mirror _all_ attributes in its "dictionary" content. This way one can choose to use either `obj.key` or `obj["key"]`.

Interesting idea :)

I'd like to point out a drawback to this approach: if the client mutates the Dict subclass after calling `to_object`, e.g. by calling `obj.key = new_value` or `obj["key"] = new_value` or `del obj.key` or `del obj["key"]`, then the dict keys of `obj` will no longer be "in sync" with the attributes of `obj` (i.e. we might get `obj.key != obj["key"]`).

Reply:
> I'd like to point out a potential drawback to this approach: if the client mutates the Dict subclass after calling `to_object`, e.g. by calling `obj.key = new_value` or `obj["key"] = new_value` or `del obj.key` or `del obj["key"]`, then the dict keys of `obj` will no longer be "in sync" with the attributes of `obj` (i.e. we might get `obj.key != obj["key"]`).

Yeah, I also thought it may not play too well with modifiying the object afterwards, but I think there are similar issues with other approaches as well. Even if initially attributes aren't mirrored, someone may call `obj.key = v` or `obj["key"] = v` thinking they are overriding their current value, while it may actually create a distinct entry (in the first case if `"key"` was actually in the dict and not an attribute, and vice-versa in the second case). I doubt we can fully avoid this kind of problem without messing with the dataclass internals (which would go against the spirit of this PR, which is to obtain a "normal" dataclass instance).

Reply:
> I doubt we can fully avoid this kind of problem without messing with the dataclass internals (which would go against the spirit of this PR, which is to obtain a "normal" dataclass instance).

Agreed.

Reply:
My opinion is that, once the data has been "exported" from OmegaConf using `to_container`/`to_object`, the responsibility for using that data in a correct way should fall on the client.

Reply:
Just to recap, I'm fine with either of these options for dataclasses inheriting from Dict:
- put everything as attributes (nothing in the dict)
- mirror everything in the dict

But the current hybrid approach seems a bit more confusing and dangerous (e.g., someone could start from a regular structured config and extend it with `open_dict()`, then later decide to make it inherit from Dict instead for convenience, and suddenly the behavior might change unexpectedly)

Reply:
> But the current hybrid approach seems a bit more confusing and dangerous (e.g., someone could start from a regular structured config and extend it with `open_dict()`, then later decide to make it inherit from Dict instead for convenience, and suddenly the behavior might change unexpectedly)

Hmm, I see your point...

Reply:
> Just to recap, I'm fine with either of these options for dataclasses inheriting from Dict:
> 
>     * put everything as attributes (nothing in the dict)
> 
>     * mirror everything in the dict

@omry, what do you think?

Reply:
Extending from Dict in Structured Configs was always is a way to mark the object as open to additions. Not a way to declare that the object should contain an actual dict fields.

I would go with the simpler solution of putting everything as attributes and ignore the fact that there the instantiated object can actually be used as a dictionary as well.

An alternative I am somewhat okay with is to put extra fields added to a Structured config extending Dict into into the inherited dictionary (__setitem__), and as attributes otherwise, but this is really getting subtle and for simplicity I prefer the first option.

I don't like replicating everything to both because it's confusing as hell and can also enable the same field to have two different values.

Bottom line, let's put it all into attributes and not treat dataclasses extending Dicts any differently.

Reply:
@Jasha10, you didn't address this one, right?

Reply:
> @Jasha10, you didn't address this one, right?

It was not addressed when you wrote that; it is now addressed (in https://github.com/omry/omegaconf/pull/502/commits/3fef7f0563bb1b304bf625e0a54a5dd42b3f9cfd).